app/src/main/contrib/components/Responsibilities/BurdenEstimates/UploadBurdenEstimatesForm.tsx

```diff
@@ -1,86 +1,345 @@
-                <h5>Second step: upload a CSV containing your central estimates</h5>
-                <PopulateEstimatesForm groupId={groupId} touchstoneId={touchstoneId} scenarioId={scenarioId} setId={estimateSet.id}/>
+    resetUploadState = (removeFile: Boolean) => {
+        this.setState({
+            ...initialUploadState,
+            metadata: {...this.state.metadata},
+            file: removeFile ? null : this.state.file,
+            fileValidationResult: removeFile ? null : this.state.fileValidationResult
+        })
+    };
+
+    componentDidMount = () => {
+
+        const settings: ConfigurationHash = {
+            ...baseSettings,
+            target: (params) => `${this.props.url}?${params.join("&")}`,
+            minFileSizeErrorCallback: () => {
+                this.props.resetPopulateState();
+                this.setState({
+                    ...initialUploadState,
+                    metadata: {...this.state.metadata},
+                    uploadErrors: [{
+                        code: "e",
+                        message: "This file is empty."
+                    }]
+                })
+            }
+        };
+
+        this.uploadClient = this.props.createUploadClient(settings);
+        this.uploadClient.assignBrowse(this.uploaderElement, false);
+
+        this.uploadClient.on('fileAdded', (file: ResumableFile) => {
+
+            this.props.resetPopulateState();
+            const validationResult = checkFileExtensionIsCSV(file.fileName);
+
+            this.setState({
+                ...initialUploadState,
+                metadata: {...this.state.metadata},
+                fileValidationResult: validationResult,
+                file: file
+            });
+        });
+
+        this.uploadClient.on('progress', () => {
+
+            this.setState({
+                isUploading: true,
+                progress: roundToOneDecimalPlace(this.uploadClient.progress() * 100)
+            });
+        });
+
+        this.uploadClient.on('fileSuccess', () => {
+
+            this.resetUploadState(false);
+            this.props.createBurdenEstimateSet(this.state.metadata);
+            this.props.populateEstimateSet(this.props.uploadToken);
+        });
+
+        this.uploadClient.on('fileError', (file: ResumableFile, error: string) => {
+
+            let uploadErrors: ErrorInfo[] = [];
+            try {
+                const result = JSON.parse(error) as Result;
+                uploadErrors = result.errors
+            } catch (e) {
+                uploadErrors.push({code: "error", message: "Error contacting server"})
+            }
+            this.setState({
+                uploadErrors: uploadErrors,
+                isUploading: false,
+                progress: 0
+            });
+        });
+    };
+
+    startUpload = () => {
+
+        this.props.resetPopulateState();
+        this.props.createBurdenEstimateSet(this.state.metadata);
+
+        const tryUpload = setInterval(() => {
```

Comment:
This method now first creates the burden estimate set (which will populate the setId and upload token in the store state) and then waits for those props to exist before starting the file upload. It's slightly ridiculous because really this is a synchronous process that is being forced into an async one by redux, but it seemed like the more minimal refactor. Alternatively we could refactor this component to manage all it's own state and avoid redux altogether...

Reply:
Do you need to do this with setInterval? Could mapStateToProps check if the burden estimate set has just been created and is ready for data (if there's a populatingSetId in the state?), and componentDidUpdate initiate the upload there, clearing populatingSetId immediately after?

Reply:
Omg yes you're right, that would be basically the same logic. Thank you!